pkg/processor/runtime/rpc/connection/connectionallocator.go

```diff
@@ -54,96 +54,180 @@ func NewConnectionAllocator(abstractConnectionManager *AbstractConnectionManager
 	return nil
-
 }
 
 func (ca *ConnectionAllocator) Stop() error {
-	var wg sync.WaitGroup
+	ca.SetStatus(status.Stopping)
 
-	for _, eventConnection := range ca.allocator.GetObjects() {
-		connection := eventConnection.(Connection)
-		wg.Add(1)
-
-		go func() {
-			defer wg.Done()
-			if err := connection.Stop(); err != nil {
-				ca.Logger.WarnWith("Failed to close connection", "error", err)
-			}
-		}()
-	}
-
-	wg.Wait()
+	err := ca.stopAllocator()
 	ca.stopControlMessageSocket()
-	return nil
+	ca.Logger.DebugWith("Stopped connection allocator",
+		"wrapperPID", ca.pid)
+
+	ca.SetStatus(status.Stopped)
+	return err
 }
 
 func (ca *ConnectionAllocator) Allocate(duration time.Duration) (eventprocessor.EventProcessor, error) {
-	// TODO: support multiple connections
 	return ca.allocator.Allocate(duration)
 }
 
+func (ca *ConnectionAllocator) Release(connection eventprocessor.EventProcessor) {
+	// if the processor requires restart, recreate the connection
+	if connection.GetStatus() == status.RestartRequired {
+		var err error
+		var newConnection []*Connection
+		// Retry connection creation up to 3 times
+	loop:
+		for i := 0; i < 3; i++ {
+			// only if the connection is in restartRequired status, try to recreate it
+			// yes, that's a double check, but it's required to cover cases where connection allocator is stopping and
+			// it will close all the connections
+			currentStatus := connection.GetStatus()
+			switch currentStatus {
+			case status.RestartRequired:
+				newConnection, err = ca.createConnections(1)
+				if err == nil {
+					break loop
+				}
+			case status.Stopped, status.Stopping:
+				// if stopped or stopping, do not release the connection
+				// it means that allocator is stopping
+				return
+			default:
+				// currently this is not really possible that status will become ready without this flow
+				// just for future
+				ca.allocator.Release(connection)
+				return
+			}
+		}
+
+		// If still failing after retries, log the error and set status to not ready (it will signal to wrapper that restart is needed)
+		if err != nil || len(newConnection) == 0 {
+			// TODO: add a background check which checks connection manager status and restarts wrapper if needed
+			// it's only added to timeout.go which checks for event timeouts
+			// for now it's fine, however if we want to add more cases when restart is needed, then it should be done separately
+			// only run if status is ready, otherwise ca is already stopping or marked as restart required
+			if ca.GetStatus() == status.Ready {
+				ca.SetStatus(status.RestartRequired)
+
+				// different log depending on the error
+				if err != nil {
+					ca.Logger.WarnWith("Failed to recreate connection after retries, wrapper restart required",
+						"rootCause", errors.RootCause(err).Error())
+				} else {
+					ca.Logger.WarnWith("Failed to recreate connection after retries, wrapper restart required",
+						"reason", "no connection created, empty slice")
+				}
+			}
```

Comment:
```suggestion
				// different log depending on the error
				reason := "no connection created, empty slice"
				if err != nil {
					reason = errors.RootCause(err).Error()
				}
				ca.Logger.WarnWith("Failed to recreate connection after retries, wrapper restart required",
						"reason",reason)
			}
```

Reply:
@TomerShor what is the point? question of taste imo and my option avoids creation of new variable

Reply:
Yes, it's only a matter of style, and to avoid duplicating the log message.